conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosGovernanceOperations.scala

```diff
@@ -0,0 +1,203 @@
+   * extract the numbers returning a collector object.
+   */
+  private def fillAggregates(
+      proposalsBlocks: Map[Block, ProtocolId],
+      listingsPerLevel: Map[Int, List[Voting.BakerRolls]],
+      ballots: Map[Block, List[Voting.Ballot]],
+      ballotCounts: Map[Block, (BallotsPerCycle, BallotsPerLevel)],
+      proposalProtocolCounts: Map[Block, Map[ProtocolId, Int]] //comes from individual operations on the block
+  ): List[GovernanceAggregate] =
+    proposalsBlocks.toList.flatMap {
+      case (block, proposal) =>
+        val listing = listingsPerLevel.get(block.data.header.level).getOrElse(List.empty)
+        val prevListings = listingsPerLevel.get(block.data.header.level - 1).getOrElse(List.empty)
+        val listingByBlock = listing.diff(prevListings)
+        val ballot = ballots.getOrElse(block, List.empty)
+        val ballotCountPerCycle = ballotCounts.get(block).map(_._1.counts)
+        val ballotCountPerLevel = ballotCounts.get(block).map(_._2.counts)
+        val proposalCounts = proposalProtocolCounts.getOrElse(block, Map.empty).toList
+
+        val allRolls = countRolls(listing, ballot)
+        val levelRolls = countRolls(listingByBlock, ballot)
+
+        /* Here we collect a row for the block being considered
+         * to get voting data for the periods with a specific proposal
+         * under scrutinee: testing vote, testing, promotion.
+         * In addition, we have many rows appended that comes
+         * from operations during the proposal period, contained
+         * in the block, which is now assumed to have no current proposal.
+         * The previous considerations would make it impossible to have
+         * the same proposal protocol and block hash for the two kind
+         * of entries just described. We only have one or the other, by
+         * construction, as the chain would reject any proposal with the
+         * same protocol as the one under voting during the ballots phases.
+         */
+        block.data.metadata match {
+          case metadata: BlockHeaderMetadata =>
+            GovernanceAggregate(
+              block.data.hash,
+              metadata,
+              Some(proposal),
+              allRolls,
+              levelRolls,
+              ballotCountPerCycle,
+              ballotCountPerLevel
+            ) ::
+                proposalCounts.map { //these come from all individual proposal operations during the proposal period
+                  case (proposalProtocol, count) =>
+                    GovernanceAggregate(
+                      block.data.hash,
+                      metadata.copy(voting_period_kind = VotingPeriod.proposal), //we know these are from operations
+                      Some(proposalProtocol),
+                      allRolls,
+                      levelRolls,
+                      Some(Voting.BallotCounts(count, 0, 0)),
+                      ballotCountPerLevel
+                    )
```

Comment:
Is there a way to express the cons logic a bit more elegantly, e.g. cons two named variables?

Reply:
Will take care of that, I agree that it's ugly